### PR TITLE
initial ipv6 implementation

### PR DESCRIPTION
**Pull request by baude**
enable ipv6 connection handling

Signed-off-by: Brent Baude <bbaude@redhat.com>

### CODE REVIEW

File: src/commands/setup.rs [hunk 1]
```diff
@@ -77,29 +81,36 @@ impl Setup {
...
                             let spf = SetupPortForward {
-                                net: (*network).clone(),
+                                net: network.clone(),
                                 container_id: network_options.container_id.clone(),
                                 port_mappings: i,
-                                network_name: (*net_name).clone(),
+                                network_name: net_name,
                                 network_hash_name: id_network_hash,
-                                options: (*per_network_opts).clone(),
+                                container_ip: container_ips[idx],
```
Comment (Luap99) [inline src/commands/setup.rs:113]: The loop order does not make sense! You use the index from the networks loop but the static_ips are per network, len(static_ips) == len(subnets). What you have to do for your port forwarding logic is loop over the static ips and take the first ipv4 (if there is one) and the first ipv6 (if there is one). If there are only ipv4 you only forward to the first ipv4 not all ips. The same for ipv6.

The same for the teardown logic

File: src/firewall/iptables.rs [hunk 2]
```diff
@@ -189,113 +192,101 @@ impl firewall::FirewallDriver for IptablesDriver {
...
-                container_ip.to_string(),
+                setup_portfw.container_ip.to_string(),
                 comment_network_cid
             ),
         )?;
 
         // FOR EACH PORT
-        for i in setup_portfw.port_mappings {
+        for i in setup_portfw.port_mappings.clone() {
+            error!("ipv6 is {}", is_ipv6);
```
Comment (Luap99) [inline src/firewall/iptables.rs:243]: This should be at least debug

Comment (baude) [reply src/firewall/iptables.rs:243]: of course, this is all just here until we confirm it works.  then i will do more code comments and cleanups

**Comment by Luap99** [comment]
For ipv6 we have to set `/proc/sys/net/ipv6/conf/<brname>/accept_dad` and `/proc/sys/net/ipv6/conf/<brname>/accept_ra` to 0. I also think the same systcls are needed for the host veth interface.

**Comment by baude** [comment]
please review and merge

File: src/network/core_utils.rs [hunk 3]
```diff
@@ -715,7 +715,18 @@ impl CoreUtils {
                 ))
             }
         }
-
+        // TODO, this should only be done for IPv6
```
Comment (flouthoc) [inline src/network/core_utils.rs:718]: Why is this inside `configure_bridge`.

File: src/network/core_utils.rs [hunk 4]
```diff
@@ -911,7 +922,16 @@ impl CoreUtils {
                 ))
             }
         }
-
+        let k = format!("/proc/sys/net/ipv6/conf/{}/enhanced_dad", &host_veth);
```
Comment (flouthoc) [inline src/network/core_utils.rs:925]: Same I dont think we should do a `sysctl` while configuring `veth`

File: src/network/core_utils.rs (hunk 3 above)
Comment (flouthoc) [reply src/network/core_utils.rs:718]: I think we should have a different function for this and call `sysctl` in `core.rs`

File: src/network/core_utils.rs (hunk 4 above)
Comment (flouthoc) [reply src/network/core_utils.rs:925]: I think we should have a different function for this and call `sysctl` in `core.rs`

**Comment by flouthoc** [comment]
@baude Above two point looks like a blocker to me. Netlink functions shouldn't be configuring `sysctls` internally.

**Comment by flouthoc** [comment]
Other than these two points LGTM

File: src/network/core_utils.rs (hunk 3 above)
Comment (flouthoc) [reply src/network/core_utils.rs:718]: Actually I'd also like to request why is this not possible in `setup.rs` or with firewall where other `sysctls` are made if `driver` is `bridge`.

File: src/network/core_utils.rs (hunk 4 above)
Comment (flouthoc) [reply src/network/core_utils.rs:925]: After checking this , we already have `host_veth` at higher level so this better in `core.rs` where we are calling function to configure bridge.

File: src/network/core_utils.rs (hunk 3 above)
Comment (Luap99) [reply src/network/core_utils.rs:718]: Yes the systctl are part of the bridge setup, I am not sure if need to set them for other drivers. Also it is important the the systctl is set before the interface is set up/get the ip assigned.

File: src/network/core_utils.rs (hunk 4 above)
Comment (Luap99) [inline src/network/core_utils.rs:925]: This is the wrong sysctl, you need to set `accept_dad` not `enhanced_dad`

Comment (Luap99) [reply src/network/core_utils.rs:925]: Also you have set both sysctls for the bridge interface as well not only the veth

**Pull request merged**